packages/mui-material/src/OverridableComponentAugmentation.ts

```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
+import { DistributiveOmit } from '@mui/types';
+
+declare module '@mui/material/OverridableComponent' {
+  /**
+   * A component whose root component can be controlled via a `component` prop.
+   *
+   * Adjusts valid props based on the type of `component`.
+   */
+  interface OverridableComponent<M extends OverridableTypeMap> {
+    <C extends React.ElementType>(
+      props: {
+        /**
+         * The component used for the root node.
+         * Either a string to use a HTML element or a component.
+         */
+        component: C;
+      } & OverridePropsVer2<M, C>,
+    ): JSX.Element;
+    (props: DefaultComponentPropsVer2<M>): JSX.Element;
+  }
+
+  /**
+   * Props of the component if `component={Component}` is used.
+   */
+  // prettier-ignore
+  type OverridePropsVer2<
+    M extends OverridableTypeMap,
+    C extends React.ElementType,
+  > = (
+    & BaseProps<M>
+    & DistributiveOmit<React.ComponentPropsWithoutRef<C>, keyof BaseProps<M>>
```

Comment:
Do you think the usage of `DistributiveOmit` here could be another bottleneck?

The reason I ask is that the React.ComponentPropsWithoutRef has a lot of keys and omitting some of the keys sounds like it requires expensive computation.

Reply:
I don't think we can get away without `DistributiveOmit` here (but I wouldn't mind being proven wrong). We need the props defined on the component (`BaseProps`) to always be present on the resulting type, even if they're not compatible with props defined on `C`.

Reply:
> Do you think the usage of DistributiveOmit here could be another bottleneck?

Nope, it was one of the first things I tried replacing, this is why I left it.